### PR TITLE
Publish artifacts use robocopy.

### PR DESCRIPTION
**Pull request by ericsciple**
- [x] Fix unit tests
- [x] E2E test

### CODE REVIEW

File: Tasks/PublishBuildArtifacts/publishbuildartifacts.ts [hunk 1]
```diff
@@ -29,13 +31,29 @@ try {
...
-        tl.cp("-Rf", path.join(pathtoPublish, "*"), artifactPath);
-            
-        // add artifactlocation to ##vso command's properties for back compat of old Xplat agent
-        data["artifactlocation"] = targetPath;
-        tl.command("artifact.associate", data, targetPath);
+
+        if (os.platform() == 'win32') {
+            var robocopy: tr.ToolRunner = new tr.ToolRunner("robocopy");
+            robocopy.arg("/E"); // copy subdirectories, including Empty ones.
+            robocopy.arg("/R:3"); // number of Retries on failed copies
```
Comment (ericsciple) [inline Tasks/PublishBuildArtifacts/publishbuildartifacts.ts]: Default retries is 1 million :)

File: Tasks/PublishBuildArtifacts/publishbuildartifacts.ts [hunk 2]
```diff
@@ -29,13 +31,29 @@ try {
...
+
         var artifactPath: string = path.join(targetPath, artifactName);
         tl.mkdirP(artifactPath);
-        tl.cp("-Rf", path.join(pathtoPublish, "*"), artifactPath);
-            
-        // add artifactlocation to ##vso command's properties for back compat of old Xplat agent
-        data["artifactlocation"] = targetPath;
-        tl.command("artifact.associate", data, targetPath);
+
+        if (os.platform() == 'win32') {
```
Comment (ericsciple) [inline Tasks/PublishBuildArtifacts/publishbuildartifacts.ts]: Check if any difference for hidden directories.

File: Tasks/PublishBuildArtifacts/publishbuildartifacts.ts [hunk 3]
```diff
@@ -29,13 +31,29 @@ try {
...
+            robocopy.arg("*"); // file
+            var result: tr.IExecResult = robocopy.execSync();
+            if (result.code >= 8) {
+                tl.setResult(tl.TaskResult.Failed, tl.loc('RobocopyFailed', result.code))
+            }
+
+            tl.command("artifact.associate", data, targetPath);
+        }
+        else {
+            tl.cp("-Rf", path.join(pathtoPublish, "*"), artifactPath);
```
Comment (ericsciple) [inline Tasks/PublishBuildArtifacts/publishbuildartifacts.ts]: Print warning?

File: Tasks/PublishBuildArtifacts/publishbuildartifacts.ts [hunk 4]
```diff
@@ -29,12 +29,33 @@ try {
...
+
+        if (os.platform() == 'win32') {
+            let robocopy: tr.ToolRunner = new tr.ToolRunner('robocopy');
+            robocopy.arg('/E'); // copy subdirectories, including Empty ones.
+            robocopy.arg('/NP'); // No Progress - don't display percentage copied.
+            robocopy.arg('/R:3'); // number of Retries on failed copies
+            robocopy.arg(pathtoPublish); // source
+            robocopy.arg(artifactPath); // destination
+            robocopy.arg('*'); // file
+            let result: tr.IExecResult = robocopy.execSync();
```
Comment (ericsciple) [inline Tasks/PublishBuildArtifacts/publishbuildartifacts.ts]: pass ignoreReturnCode = true

**Comment by ericsciple** [comment]
I'm still working on adding more unit tests for negative cases.

File: Tasks/PublishBuildArtifacts/publishbuildartifacts.ts (hunk 2 above)
Comment (ericsciple) [inline Tasks/PublishBuildArtifacts/publishbuildartifacts.ts]: output encoding is different.
need to investigate file acl differences.

**Comment by bryanmacfarlane** [comment]
When done, let's make sure we port to 105 branch to get RC2 coverage and feedback.  We should try and do that today to avoid ask mode process.
